### PR TITLE
Adding components and refactoring of schedulers

### PR DESCRIPTION
**Pull request by ericspod**
Addresses #280.

The objective here is define a base class for all schedulers that we can use for reducing duplication of code and type checking. 

One thing added is a way of adding new noise schedule functions by adding them to a component store object. This version adds the cosine schedule but others can be added by users in their own scripts.

### CODE REVIEW

**Comment by marksgraham** [comment]
Hi @ericspod 
Sorry for taking so long, now GTC+MICCAI are out the way I'll be more responsive. 

> We could do that easily enough, my concern in MONAI and elsewhere is to make explicit what functions are available and give them useful names. We have enumerations for int and string constants in MONAI for this reason as opposed to the way other libraries take of just stating valid values in the docstring. Our layer factories make it easy to choose a layer by name with the dimensionality being parameter, but they also gives clear names to the concepts they store. [Explicit is better than implicit, but simple is better than complex](https://peps.python.org/pep-0020/) so the question is if the tradeoff makes sense here.

OK, then I think the tradeoff does make sense here! I guess the next step is to update all the other schedulers? In which case I'll hold off carefully reviewing the latest batch of changes until thats done?

**Comment by ericspod** [comment]
Hi @marksgraham, I have things updated finally if you want to take a look.

**Comment by marksgraham** [comment]
Hi @ericspod 
I've taken a look and looks good - shall I hold off until you've fixed the schedule to do a formal review?
NB that the diffusion tutorials will have to be updated. Because you've kept the defaults the same it should be sufficient to just update the arguments, I don't think youll need to rerun all the notebooks.

File: generative/networks/schedulers/scheduler.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        scheduler = DDPMScheduler(1000, "my_beta_schedule")
+        
+    Args:
+        num_train_timesteps: number of diffusion steps used to train the model.
+        schedule: member of NoiseSchedules,
+            a named function returning the beta tensor or (betas, alphas, alphas_cumprod) triple
+        schedule_args: arguments to pass to the schedule function
+    """
+
+    def __init__(self, num_train_timesteps: int = 1000, schedule: str = "linear", **schedule_args) -> None:
```
Comment (Warvito) [inline generative/networks/schedulers/scheduler.py:141]: typo `linear_beta`

**Review by Warvito — CHANGES_REQUESTED** [review]

File: generative/networks/schedulers/scheduler.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+    Returns:
+        betas: beta schedule tensor
+    """
+    betas = torch.linspace(-sig_range, sig_range, num_train_timesteps)
+    return torch.sigmoid(betas) * (beta_end - beta_start) + beta_start
+
+
+@NoiseSchedules.add_def("cosine_beta", "Cosine beta schedule")
+def _cosine_beta(num_train_timesteps, s=8e-3):
```
[the remark this answers is not in the document]
Comment (ericspod) [reply generative/networks/schedulers/scheduler.py:96]: `s` is referenced in the paper so I'd leave it here to be consistent with that.

**Comment by ericspod** [comment]
@Warvito Updates from comments added!

File: generative/networks/schedulers/ddpm.py [hunk 3]
```diff
@@ -33,70 +33,71 @@
...
-        self.prediction_type = prediction_type
+        if variance_type not in DDPMVarianceType.__members__.values():
+            raise ValueError("Argument `variance_type` must be a member of `DDPMVarianceType`")
 
-        self.num_train_timesteps = num_train_timesteps
-        self.alphas = 1.0 - self.betas
-        self.alphas_cumprod = torch.cumprod(self.alphas, dim=0)
-        self.one = torch.tensor(1.0)
+        if prediction_type not in DDPMPredictionType.__members__.values():
+            raise ValueError("Argument `prediction_type` must be a member of `DDPMPRedictionType`")
```
Comment (marksgraham) [inline generative/networks/schedulers/ddpm.py:96]:
```suggestion
            raise ValueError("Argument `prediction_type` must be a member of `DDPMPredictionType`")
```

**Review by marksgraham — APPROVED** [review]
Hi @ericspod 

This looks good. Just need to run `./runtests.sh --autofix` as it picks up some formatting corrections

**Comment by Warvito** [comment]
I am getting these flake8 errors when running `./runtests.sh -f -u --net`

```
flake8
6.0.0 (flake8-bugbear: 22.12.6, flake8-comprehensions: 3.10.1, flake8-executable: 2.1.2, mccabe: 0.7.0, pep8-naming: 0.13.3, pycodestyle: 2.10.0, pyflakes: 3.0.1) CPython 3.10.6 on Linux
/media/walter/Storage/Projects/GenerativeModels/generative/networks/schedulers/ddim.py:89:30: F541 f-string is missing placeholders
/media/walter/Storage/Projects/GenerativeModels/generative/networks/schedulers/pndm.py:92:30: F541 f-string is missing placeholders
/media/walter/Storage/Projects/GenerativeModels/generative/networks/schedulers/scheduler.py:35:1: F401 'typing.Callable' imported but unused
/media/walter/Storage/Projects/GenerativeModels/generative/networks/schedulers/scheduler.py:37:1: F401 'numpy as np' imported but unused
/media/walter/Storage/Projects/GenerativeModels/generative/utils/__init__.py:16:1: F403 'from .misc import *' used; unable to detect undefined names
2     F401 'typing.Callable' imported but unused
1     F403 'from .misc import *' used; unable to detect undefined names
2     F541 f-string is missing placeholders
```

**Comment by ericspod** [comment]
@Warvito I've made those fixes but my version of the tools brought up some other issues, different versions of flake8 keep coming up with new issues not seem before.

**Comment by marksgraham** [comment]
Do we need to settle on a version of flake8 to use then?

**Comment by ericspod** [comment]
We should always use the current version I think, it just means we'll encounter errors unrelated to our own changes on occasion.

**Review by Warvito — APPROVED** [review]
